onnxscript/rewriter/pattern.py

```diff
@@ -755,47 +875,66 @@ def __init__(
     @property
     def tag_var(self) -> str | None:
         """Returns the tag variable associated with the OrValue pattern."""
         return self._tag_var
 
-    def clone(self, node_map: dict[NodePattern, NodePattern]) -> OrValue:
-        return OrValue(
+    def clone(self, node_map: dict[NodePattern, NodePattern]) -> _BacktrackingOr:
+        return _BacktrackingOr(
             [v.clone(node_map) for v in self._values],
             self.name,
             self._tag_var,
             self._tag_values,
         )
 
-    def get_pattern(self, value: ir.Value) -> tuple[Any, ValuePattern]:
-        """Returns the pattern that should be tried for the given value."""
-        producer = value.producer()
-        if producer is not None:
+
+def OrValue(
+    values: Sequence[ValuePattern],
+    name: str | None = None,
+    tag_var: str | None = None,
+    tag_values: Sequence[Any] | None = None,
+) -> ValuePattern:
+    """
+    Creates an OR pattern.
+
+    Args:
+        values: A sequence of value patterns to match against.
+        name: An optional variable name for the pattern. Defaults to None. If present,
+            this name will be bound to the value matched by the pattern.
+        tag_var: An optional variable name for the tag. Defaults to None. If present,
+            it will be bound to a value (from tag_values) indicating which alternative was matched.
+        tag_values: An optional sequence of values to bind to the tag_var. Defaults to None.
+            If present, the length of tag_values must match the number of alternatives in values.
+            In a successful match, tag-var will be bound to the i-th value in tag_values if the i-th
+            alternative pattern matched. If omitted, the default value of (0, 1, 2, ...) will be used.
+    """
+    if tag_values is not None:
+        if tag_var is None:
+            raise ValueError("tag_var must be specified if tag_values is provided.")
+        if len(tag_values) != len(values):
+            raise ValueError(
+                "tag_values must have the same length as the number of alternatives."
+            )
+    else:
+        tag_values = tuple(range(len(values)))
+
+    def make_op_id_or_pattern() -> _OpIdDispatchOr | None:
+        mapping: dict[ir.OperatorIdentifier, tuple[Any, NodeOutputPattern]] = {}
+        for i, alternative in enumerate(values):
+            if not isinstance(alternative, NodeOutputPattern):
+                return None
+            producer = alternative.producer()
             id = producer.op_identifier()
-            if id is not None and id in self._op_to_pattern:
-                return self._op_to_pattern[id]
-        return self._default_pattern
+            if id is None or id in mapping:
+                return None
+            mapping[id] = (tag_values[i], alternative)
+        return _OpIdDispatchOr(mapping, name, tag_var)
+
+    optimized_pattern = make_op_id_or_pattern()
+    return optimized_pattern or _BacktrackingOr(
```

Comment:
Given _BacktrackingOr is sort of a fallback mechanism here, could this cause any speedup issues where we have a lot of OrValues (that fallback to backtracking) used in the pattern?

Reply:
The alternative would be to convert them into multiple rules, which would take even more time. That is what we need to compare against. If it is not fast enough, well, then we need to design an even more efficient algorithm.